### PR TITLE
Fix: database execution is not using transaction functions

### LINKED ISSUE #1688: Database execution is not using transaction functions
**Describe the bug**

Raised by a customer, details in Trello.

### PR DESCRIPTION
**Pull request by darrellwarde**
# Description

#953 inadvertently removed a useful feature provided by using the driver's transaction function, which retry a query a number of times in the event of execution failure. This PR reintroduced this functionality, with a refactor breaking out execution code into an `Executor` class.

This should be reviewed, merged and released with fairly high priority.

# Issue

Closes #1688.

### CODE REVIEW

File: packages/graphql/src/classes/Executor.ts [hunk 1]
```diff
@@ -0,0 +1,215 @@
+/*
```
Comment (tbwiss) [inline packages/graphql/src/classes/Executor.ts:1]: question: Do we have to make any updates/additions to the documentation?

File: packages/graphql/src/classes/Executor.ts [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+                const result = await this.sessionRun(query, parameters, defaultAccessMode, session);
+                await session.close();
+                return result;
+            }
+
+            if (isSessionLike(this.executionContext)) {
+                return await this.sessionRun(query, parameters, defaultAccessMode, this.executionContext);
+            }
+
+            return await this.transactionRun(query, parameters, this.executionContext);
```
Comment (tbwiss) [inline packages/graphql/src/classes/Executor.ts:116]: question: what about calling `transaction.commit()` after `transaction.run()`? Is that still needed or is that a thing of the past?
(this maybe shows more that I'm not quite sure how it all works but I thought of bringing it up anyway)

File: packages/graphql/src/classes/Executor.ts [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+export type ExecutorConstructorParam = {
+    executionContext: ExecutionContext;
+    auth: AuthContext;
+    queryOptions?: CypherQueryOptions;
+    database?: string;
+    bookmarks?: string | string[];
+};
+
+export class Executor {
```
Comment (tbwiss) [inline packages/graphql/src/classes/Executor.ts:82]: Nice refactor!

File: packages/graphql/src/classes/Executor.ts (hunk 1 above)
Comment (darrellwarde) [reply packages/graphql/src/classes/Executor.ts:1]: We do not - these changes were documented in #953, and this is just fixing some of the underlying implementation. 🙂

File: packages/graphql/src/classes/Executor.ts (hunk 2 above)
Comment (darrellwarde) [reply packages/graphql/src/classes/Executor.ts:116]: So this is a case of when a user has passed us their own transaction, which they want to be in control of - so they will have to perform the commit after the execution is complete.

File: packages/graphql/src/classes/Executor.ts [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        return {
+            metadata: {
+                app,
+                type: "user-transpiled",
+            },
+        };
+    }
+
+    private async sessionRun(query: string, parameters, defaultAccessMode, session): Promise<QueryResult> {
+        const result = await session[`${defaultAccessMode.toLowerCase()}Transaction`](
```
Comment (angrykoala) [inline packages/graphql/src/classes/Executor.ts:193]: **nitpick, non-blocking** for readability purposes, maybe extract `${defaultAccessMode.toLowerCase()}Transaction` into a separate varaible "transactionName" or something like that

**Review by angrykoala — APPROVED** [review]
Nice refactor

File: packages/graphql/src/classes/Executor.ts (hunk 4 above)
Comment (darrellwarde) [reply packages/graphql/src/classes/Executor.ts]: Good shout!

**Pull request merged**